### PR TITLE
[ticket/15695] Fix gen_rand_string returning less characters than expected

### PR DESCRIPTION
**Pull request by rubencm**
PHPBB3-15695

Checklist:

- [x] Correct branch: master for new features; 3.2.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/dev/master/development/coding_guidelines.html) and [3.2.x](https://area51.phpbb.com/docs/dev/3.2.x/development/coding_guidelines.html)
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.2.x/development/git.html)

Tracker ticket (set the ticket ID to **your ticket ID**):

https://tracker.phpbb.com/browse/PHPBB3-15695

### CODE REVIEW

File: phpBB/includes/functions.php [hunk 1]
```diff
@@ -70,10 +70,18 @@ function set_var(&$result, $var, $type, $multibyte = false)
 *
 * @return string
 */
-function gen_rand_string($num_chars = 8)
+function gen_rand_string($num_chars = 8, $characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890')
 {
-	// [a, z] + [0, 9] = 36
-	return substr(strtoupper(base_convert(bin2hex(random_bytes($num_chars)), 16, 36)), 0, $num_chars);
```
Comment (marc1706) [inline phpBB/includes/functions.php]: Why not just change this to `substr(strtoupper(base_convert(bin2hex(random_bytes($num_chars + 1)), 16, 36)), 0, $num_chars);`? Should work just fine and you'll not be using something that should be actually random.

Comment (rubencm) [reply phpBB/includes/functions.php]: My version is better, the old one fails with big quantities of characters too

Comment (rubencm) [reply phpBB/includes/functions.php]: The expected length fails with 65+ characters with your modification, also base_convert function starts to fail with 127+ characters

File: phpBB/includes/functions.php [hunk 2]
```diff
@@ -66,21 +66,23 @@ function set_var(&$result, $var, $type, $multibyte = false)
...
+* @param int $num_chars Length of random string, defaults to 8.
+* This number should be less or equal than 64.
 *
 * @return string
 */
 function gen_rand_string($num_chars = 8)
 {
 	// [a, z] + [0, 9] = 36
-	return substr(strtoupper(base_convert(bin2hex(random_bytes($num_chars)), 16, 36)), 0, $num_chars);
+	return substr(strtoupper(base_convert(bin2hex(random_bytes($num_chars+1)), 16, 36)), 0, $num_chars);
```
Comment (marc1706) [inline phpBB/includes/functions.php]: Add spaces to comply with coding guidelines: `$num_chars + 1`

Comment (marc1706) [reply phpBB/includes/functions.php]: Please also add this to gen_rand_string_friendly() (the + 1).

**Pull request merged**